### PR TITLE
Binding to graph API V1

### PR DESCRIPTION
**Pull request by hrandika**
## Automation tests
 - Unit tests 
   > None
 - Integration tests
   > Test are commented out,Update test with details with client_id and client_secret

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? no
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+            ClientInformation appInfo = this.getClientInformation(oauthAppInfo);
+            ClientInformation app;
+
+            try {
+                app = appClient.createApplication(appInfo);
+                if (app != null) {
+                    PasswordInfo pInfo = this.setPassword(app.getId());
+                    app.setClientSecret(pInfo.getSecret());
+                    return getOAuthApplicationInfo(app);
+                }
```
Comment (malinthaprasan) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:89]: Is the else block of this "if" an error situation?
Should we handle this case too?

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    }
+
+    @Override
+    public void deleteApplication(String clientId) throws APIManagementException {
+        ClientInformation client = getClientInformationByClientId(clientId);
+        if (client != null) {
+            try {
+                appClient.deleteApplication(client.getId());
+            } catch (KeyManagerClientException e) {
+                e.printStackTrace();
```
Comment (malinthaprasan) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:169]: remove printstack trace

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    }
+
+    private ClientInformation getClientInformationByClientId(String clientId) throws APIManagementException {
+        ClientInformation client = null;
+        try {
+            ClientInformationList list = appClient.searchByAppId(clientId);
+            if (list != null && list.getValue().size() > 0)
+                client = list.getValue().get(0);
+        } catch (KeyManagerClientException e1) {
+            e1.printStackTrace();
```
Comment (malinthaprasan) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:182]: Lets check and remove printStackTrace() from all the places

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+            }
+        }
+    }
+
+    private ClientInformation getClientInformationByClientId(String clientId) throws APIManagementException {
+        ClientInformation client = null;
+        try {
+            ClientInformationList list = appClient.searchByAppId(clientId);
+            if (list != null && list.getValue().size() > 0)
+                client = list.getValue().get(0);
```
Comment (malinthaprasan) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:180]: Do not use inline if blocks

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    protected void clearTest() throws KeyManagerClientException {
+        ClientInformationList allTestApplications = this.appClient.getAllTestApplications();
+        allTestApplications.getValue().forEach(v -> {
+            try {
+                this.appClient.deleteApplication(v.getId());
+            } catch (KeyManagerClientException e) {
+                e.printStackTrace();
+            }
+        });
+    }
```
Comment (malinthaprasan) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:215]: What is the use of this method?

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+        String consumerSecret = oAuthAppRequest.getOAuthApplicationInfo().getClientSecret();
+
+        if (StringUtils.isNotBlank(consumerKey)) {
+            OAuthApplicationInfo clientInfo = retrieveApplication(consumerKey);
+            if (clientInfo == null) {
+                String msg = "Something went wrong while getting OAuth application for given consumer key "
+                        + consumerKey;
+                log.error(msg);
+                throw new APIManagementException(msg);
+            }
```
Comment (malinthaprasan) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:234]: Let not log and throw. 
```suggestion
                throw new APIManagementException(msg);
            }
```

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADRequestInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package org.wso2.azure.client;
+
+import org.wso2.carbon.apimgt.impl.APIConstants;
+import org.wso2.carbon.apimgt.impl.recommendationmgt.AccessTokenGenerator;
```
Comment (malinthaprasan) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADRequestInterceptor.java:4]: Looking at the implementation of this class, I guess we can move this to a common location @tharindu1st ?
Getting a class to an azure ad implementation from APIM recommendation feature is confusing and also it can break easily if someone do any change to the class related to the recommendation feature.

I guess the best thing to do for now is, copying this class into this repo itself and use it.

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/TokenGenerator.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                .url(tokenEndpoint)
+                .post(formBody)
+                .build();
+        Call call = client.newCall(request);
+
+        try (Response response = call.execute()) {
+            return new Gson().fromJson(response.body().string(), AccessTokenResponse.class);
+        } catch (IOException e) {
+            // e.printStackTrace();
+            log.error(e.getMessage());
```
Comment (malinthaprasan) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/TokenGenerator.java:40]: Should we throw an exception?

File: src/main/java/org/wso2/km/ext/azure/App.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package org.wso2.km.ext.azure;
```
Comment (malinthaprasan) [inline src/main/java/org/wso2/km/ext/azure/App.java:1]: Remove unused class.
Please check and remove all unused classes and files.

File: components/azure.key.manager/pom.xml [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            <artifactId>json-simple</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.wso2.carbon.apimgt</groupId>
+            <artifactId>org.wso2.carbon.apimgt.impl</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.24</version>
```
Comment (malinthaprasan) [inline components/azure.key.manager/pom.xml:37]: Move the version to the root pom.
Please check and move all the versions like this.

File: components/azure.key.manager/pom.xml [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        <dependency>
+            <groupId>io.github.openfeign</groupId>
+            <artifactId>feign-slf4j</artifactId>
+        </dependency>
+
+        <!-- Test -->
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <version>5.9.0</version>
```
Comment (malinthaprasan) [inline components/azure.key.manager/pom.xml:65]: Move the version to the root pom

File: components/azure.key.manager/pom.xml [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>3.0.0-M7</version>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.junit.jupiter</groupId>
+                        <artifactId>junit-jupiter-engine</artifactId>
+                        <version>5.9.0</version>
+                    </dependency>
```
Comment (malinthaprasan) [inline components/azure.key.manager/pom.xml:118]: Is it possible to move these versions to the root pom too?

File: components/azure.key.manager/pom.xml [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+                </dependencies>
+            </plugin>
+        </plugins>
+    </build>
+
+</project>
```
Comment (malinthaprasan) [inline components/azure.key.manager/pom.xml:124]: Add new line at the end of file.
Please check this for all files and add a new line if its not there.

**Comment by tharindu1st** [comment]
include spotbugs and checkstyle plugins https://github.com/wso2-extensions/apim-km-wso2is/blob/master/pom.xml#L61-L104 @hrandika

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 14]
```diff
@@ -0,0 +1,332 @@
+package org.wso2.azure.client;
```
Comment (chamilaadhi) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:1]: Missing license header

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/ApplicationClient.java [hunk 15]
```diff
@@ -0,0 +1,56 @@
+package org.wso2.azure.client;
```
Comment (chamilaadhi) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/ApplicationClient.java:1]: Missing license header

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 16]
```diff
@@ -0,0 +1,332 @@
...
+        ClientInformation clientInformation = new ClientInformation();
+        clientInformation.setAppName(oauthAppInfo.getClientName());
+        String id = (String) oauthAppInfo.getParameter(AzureADConstants.OBJECT_ID);
+        if (id != null)
+            clientInformation.setId(id);
+
+        if (oauthAppInfo.getClientId() != null)
+            clientInformation.setClientId(oauthAppInfo.getClientId());
+
+        if (oauthAppInfo.getClientSecret() != null)
```
Comment (chamilaadhi) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:134]: please use the following format
```
if () {
}
```

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 17]
```diff
@@ -0,0 +1,332 @@
...
+        // Add object UUID
+        oauthAppInfo.addParameter(AzureADConstants.OBJECT_ID, appInfo.getId());
+
+        return oauthAppInfo;
+    }
+
+    private ClientInformation getClientInformation(OAuthApplicationInfo oauthAppInfo) {
+        ClientInformation clientInformation = new ClientInformation();
+        clientInformation.setAppName(oauthAppInfo.getClientName());
+        String id = (String) oauthAppInfo.getParameter(AzureADConstants.OBJECT_ID);
```
Comment (chamilaadhi) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:127]: You have done a null check for `id` in the next line. One possible scenario for the id to become null is `oauthAppInfo.getParameter(AzureADConstants.OBJECT_ID)` getting null. If that happens there will be an exception when casting to String in this line. You have to check whether that property exists before casting to string

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 18]
```diff
@@ -0,0 +1,332 @@
...
+    @Override
+    public OAuthApplicationInfo createApplication(OAuthAppRequest oauthAppRequest) throws APIManagementException {
+        OAuthApplicationInfo oauthAppInfo = oauthAppRequest.getOAuthApplicationInfo();
+
+        if (oauthAppInfo != null) {
+            ClientInformation appInfo = this.getClientInformation(oauthAppInfo);
+            ClientInformation app;
+
+            try {
+                app = appClient.createApplication(appInfo);
```
Comment (chamilaadhi) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:84]: Add debug logs to identify payload, etc

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+
+    @Override
+    public OAuthApplicationInfo updateApplication(OAuthAppRequest oauthAppRequest) throws APIManagementException {
+        OAuthApplicationInfo oauthAppInfo = oauthAppRequest.getOAuthApplicationInfo();
+        if (oauthAppInfo != null) {
+            ClientInformation appInfo = this.getClientInformation(oauthAppInfo);
+            String id = appInfo.getId();
+
+            try {
+                // Update, Client does not send a body. 204 Status only
```
Comment (chamilaadhi) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:148]: Add debug logs to identify payload, etc

File: .mvn/wrapper/maven-wrapper.properties [hunk 20]
```diff
@@ -0,0 +1,2 @@
+distributionUrl=https://repo.maven.apache.org/maven2/org/apache/maven/apache-maven/3.6.0/apache-maven-3.6.0-bin.zip
+wrapperUrl=https://repo.maven.apache.org/maven2/org/apache/maven/wrapper/maven-wrapper/3.1.0/maven-wrapper-3.1.0.jar
```
Comment (chamilaadhi) [inline .mvn/wrapper/maven-wrapper.properties:2]: Is this file needed?

File: .vscode/settings.json [hunk 21]
```diff
@@ -0,0 +1,5 @@
+{
+  "java.configuration.updateBuildConfiguration": "automatic",
+  "java.jdt.ls.vmargs": "-XX:+UseParallelGC -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xmx2G -Xms100m -javaagent:\"/home/randika/.vscode/extensions/vscjava.vscode-lombok-1.0.1/server/lombok.jar\"",
+  "workbench.colorTheme": "Solarized Light"
+}
```
Comment (chamilaadhi) [inline .vscode/settings.json:5]: do we need this file? please check whether there are any other unwanted files

File: components/azure.key.manager/pom.xml [hunk 22]
```diff
@@ -0,0 +1,124 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (madushadhanushka) [inline components/azure.key.manager/pom.xml:1]: Need to add licence headers

File: components/azure.key.manager/src/test/java/org/wso2/azure/client/AzureAdValues.java [hunk 23]
```diff
@@ -0,0 +1,10 @@
+package org.wso2.azure.client;
+
+// https://graph.microsoft.com
+// https://login.microsoftonline.com/025464ed-e523-4fb2-94fe-166720e9f0a2/federationmetadata/2007-06/federationmetadata.xml
```
Comment (madushadhanushka) [inline components/azure.key.manager/src/test/java/org/wso2/azure/client/AzureAdValues.java:4]: Shall we remove these comments if not required

File: components/azure.key.manager/src/test/java/org/wso2/azure/client/TestAzureADClient.java [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+        this.azureADClient.deleteApplication(clientId);
+    }
+
+    // @Test
+    public void itShouldGetClientInformation() throws APIManagementException,
+            InterruptedException {
+        String clientId = this.createAndGetClientId();
+
+        // this is needed since update take a bit of time to reflect.
+        Thread.sleep(30000);
```
Comment (madushadhanushka) [inline components/azure.key.manager/src/test/java/org/wso2/azure/client/TestAzureADClient.java:77]: Can we do retry instead of hard code with timeout?

File: test.sh [hunk 25]
```diff
@@ -0,0 +1,3 @@
+./mvnw clean install
+cp components/azure.key.manager/target/manager-1.0-SNAPSHOT.jar dockerfiles/apim/
+docker-compose up --build
```
Comment (madushadhanushka) [inline test.sh:3]: Lets add an empty line below

File: mvnw [hunk 26]
```diff
@@ -0,0 +1,316 @@
+#!/bin/sh
+# ----------------------------------------------------------------------------
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (madushadhanushka) [inline mvnw:3]: wrong licence headers

File: components/azure.key.manager/src/test/java/org/wso2/azure/client/TestAzureADRequestInterceptor.java [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+import org.wso2.carbon.apimgt.impl.APIConstants;
+
+import feign.RequestTemplate;
+
+@ExtendWith(MockitoExtension.class)
+public class TestAzureADRequestInterceptor {
+
+    AzureADRequestInterceptor adRequestInterceptor;
+
+    // @BeforeEach
```
Comment (madushadhanushka) [inline components/azure.key.manager/src/test/java/org/wso2/azure/client/TestAzureADRequestInterceptor.java:21]: We can remove this `@beforeeach` if not required

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java (hunk 1 above)
Comment (hrandika) [reply components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:89]: handled with an else and throw an exception

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java (hunk 5 above)
Comment (hrandika) [reply components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADClient.java:215]: Unwanted code. Used for unit testing.

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADRequestInterceptor.java (hunk 7 above)
Comment (hrandika) [reply components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADRequestInterceptor.java:4]: No action items taken to this point

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/TokenGenerator.java (hunk 8 above)
Comment (hrandika) [reply components/azure.key.manager/src/main/java/org/wso2/azure/client/TokenGenerator.java:40]: Added

File: components/azure.key.manager/pom.xml (hunk 10 above)
Comment (hrandika) [reply components/azure.key.manager/pom.xml:37]: Removed the dependency

File: components/azure.key.manager/pom.xml (hunk 11 above)
Comment (hrandika) [reply components/azure.key.manager/pom.xml:65]: Removed the dependency

File: components/azure.key.manager/pom.xml (hunk 12 above)
Comment (hrandika) [reply components/azure.key.manager/pom.xml:118]: Removed the dependency

File: .mvn/wrapper/maven-wrapper.properties (hunk 20 above)
Comment (hrandika) [reply .mvn/wrapper/maven-wrapper.properties:2]: This file is a part of mvnw https://maven.apache.org/wrapper/

File: .vscode/settings.json (hunk 21 above)
Comment (hrandika) [reply .vscode/settings.json:5]: removed

File: components/azure.key.manager/src/test/java/org/wso2/azure/client/AzureAdValues.java (hunk 23 above)
Comment (hrandika) [reply components/azure.key.manager/src/test/java/org/wso2/azure/client/AzureAdValues.java:4]: files removed since its not the best solution to test

File: components/azure.key.manager/src/test/java/org/wso2/azure/client/TestAzureADClient.java (hunk 24 above)
Comment (hrandika) [reply components/azure.key.manager/src/test/java/org/wso2/azure/client/TestAzureADClient.java:77]: files removed since its not the best solution to test

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADRequestInterceptor.java (hunk 7 above)
Comment (tharindu1st) [reply components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADRequestInterceptor.java:21]: @malinthaprasan +1

**Comment by hrandika** [comment]
PR updated with above changes

File: components/azure.key.manager/pom.xml [hunk 28]
```diff
@@ -0,0 +1,104 @@
...
+
+    <parent>
+        <groupId>org.wso2.km.ext.azure</groupId>
+        <artifactId>client</artifactId>
+        <version>1.0-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+
+    <groupId>org.wso2.km.ext.azure</groupId>
+    <artifactId>manager</artifactId>
```
Comment (tharindu1st) [inline components/azure.key.manager/pom.xml:31]: azure.key.manager

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/ApplicationClient.java [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+import org.wso2.azure.client.model.ClientInformation;
+import org.wso2.azure.client.model.ClientInformationList;
+import org.wso2.azure.client.model.PasswordInfo;
+import org.wso2.carbon.apimgt.impl.kmclient.KeyManagerClientException;
+
+import feign.Headers;
+import feign.Param;
+import feign.RequestLine;
+
+public interface ApplicationClient {
```
Comment (tharindu1st) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/ApplicationClient.java:29]: missing class comment

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADConnectorConfiguration.java [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+public class AzureADConnectorConfiguration implements KeyManagerConnectorConfiguration {
+
+    @Override
+    public String getImplementation() {
+        return AzureADClient.class.getName();
+    }
+
+    @Override
+    public String getJWTValidator() {
+        return null;
```
Comment (tharindu1st) [inline components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADConnectorConfiguration.java:40]: did we try to validate JWT generate from azure from Gateway

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADConnectorConfiguration.java (hunk 30 above)
Comment (hrandika) [reply components/azure.key.manager/src/main/java/org/wso2/azure/client/AzureADConnectorConfiguration.java:40]: Token was generated and validate manually with JWT.io

File: mvnw (hunk 26 above)
Comment (hrandika) [reply mvnw]: removed mvnw

**Comment by hrandika** [comment]
updated the PR with request changes

**Review by tharindu1st — APPROVED** [review]

**Review by tharindu1st — APPROVED** [review]
